src/ir/stack-utils.cpp

```diff
@@ -0,0 +1,272 @@
+  }
+  bool paramSuffixMatches =
+    std::equal(params.begin(),
+               params.end(),
+               sig.params.end() - params.size(),
+               [](const Type& consumed, const Type& provided) {
+                 return Type::isSubType(provided, consumed);
+               });
+  if (!paramSuffixMatches) {
+    return false;
+  }
+  bool resultSuffixMatches =
+    std::equal(results.begin(),
+               results.end(),
+               sig.results.end() - results.size(),
+               [](const Type& produced, const Type& expected) {
+                 return Type::isSubType(produced, expected);
+               });
+  if (!resultSuffixMatches) {
+    return false;
+  }
+  if (unreachable) {
+    // The unreachable can consume any additional provided params and produce
+    // any additional expected results, so we are done.
+    return true;
+  }
+  // Any additional provided params will pass through untouched, so they must be
+  // quivalent to the additional produced results.
+  return std::equal(sig.params.begin(),
+                    sig.params.end() - params.size(),
+                    sig.results.begin(),
+                    sig.results.end() - results.size(),
+                    [](const Type& produced, const Type& expected) {
+                      return Type::isSubType(produced, expected);
+                    });
+}
+
+StackSignature& StackSignature::operator+=(const StackSignature& next) {
+  assert(composes(next));
+  std::vector<Type> stack(results.begin(), results.end());
+  size_t required = next.params.size();
+  // Consume stack values according to next's parameters
+  if (stack.size() >= required) {
+    stack.resize(stack.size() - required);
+  } else {
+    if (!unreachable) {
+      // Prepend the unsatisfied params of `next` to the current params
+      size_t unsatisfied = required - stack.size();
+      std::vector<Type> newParams(next.params.begin(),
+                                  next.params.begin() + unsatisfied);
+      newParams.insert(newParams.end(), params.begin(), params.end());
+      params = Type(newParams);
+    }
+    stack.clear();
+  }
+  // Add stack values according to next's results
+  if (next.unreachable) {
+    results = next.results;
+    unreachable = true;
+  } else {
+    stack.insert(stack.end(), next.results.begin(), next.results.end());
+    results = Type(stack);
+  }
+  return *this;
+}
+
+StackSignature StackSignature::operator+(const StackSignature& next) const {
+  StackSignature sig = *this;
+  sig += next;
+  return sig;
+}
+
+StackFlow::StackFlow(Block* block) {
+  // Abstract out the logic for making a pass through the block
+  auto processBlock = [&block](auto process) {
+    for (auto* expr : block->list) {
+      process(expr, StackSignature(expr));
+    }
+    bool unreachable = block->type == Type::unreachable;
+    Type params = unreachable ? Type::none : block->type;
+    process(block, StackSignature(params, Type::none, unreachable));
```

Comment:
The `StackSignature` passed to `process` describes the values produced and consumed by the expression (in this case, `block`). The consumed values correspond to the signature's `params` and the produced values correspond 
 to the signature's `results`. From the point of view of the instructions inside a block, the end of the block "consumes" values corresponding to the block's return type and since we are only considering the block and its children, it doesn't make sense for the end of the block to "produce" any values. By setting up the signature this way, `process` can treat the end of the block just like any other instruction that consumes values.

I'll add a comment clarifying this in the code.

Reply:
> From the point of view of the instructions inside a block, the end of the block "consumes" values corresponding to the block's return type and since we are only considering the block and its children, it doesn't make sense for the end of the block to "produce" any values.

Sorry I can't still understand why the end of the block is specially treated this way. I'm not even sure what the 'end of the block' is. Do you mean an `end` instruction to a block?

For each expression within a block, the processing call is
```
process(expr, StackSignature(expr)); 
```
The first argument is an expression, and the second argument is a `StackSignature` of that expression. But this relationship breaks in the last line (and you explained the reason, but I couldn't understand why that should be that way):
```
Type params = unreachable ? Type::none : block->type;                        
process(block, StackSignature(params, Type::none, unreachable));
```
The first argument is `block`, but the second argument is not a `StackSignature` of it; it is rather a `StackSignature` with its `params` and `results` swapped.

Reply:
> a Block expression's srcs are not the instructions or expressions that come before this Block but rather within this Block, is this correct?

Yes, exactly.

> I also wonder is there a similar proxy for the start of a block? We don't have block input values now, but when we have it, it sounds consistent to have a start of block proxy too. But block itself is used for a proxy of the end of the block. What are we gonna use for the block starts?

Yes, in the future when we support blocks somehow taking stack parameters we will want to do a similar thing where we do `process(block, StackSignature{Type::none, block->params, false})` before iterating through the block. That will set up the stack in the proper initial configuration and will make the block the `src` for all the values that are on the stack at the beginning (just like how the block is the `dest` for all values left on the stack at the end now).

> I still wonder what these start and end markers are used for, though

Very generally, they could be useful for determining which values flow into or out of a block, but I haven't written any code that specifically depends on that yet. The real reason they exist is that we need to record _some_ `dest` for values that are produced inside the block but not consumed inside the block, and `block` is the best available value. For example, consider the following type inference problem where we have to determine what signature to emit for $a:

```
(func $foo (result i32 i64)
  (block $body (result i32 i64)
    (f32.const 0)
    (block $a [unreachable]
      ...
    )
    (i64.const 0)
  )
)
```

To do type inference for $a, we compute a `StackFlow` for $body, which tells us that $a consumes an f32 and produces an i32. But that i32 it produces is not consumed inside $body, so if we ignored values that are not consumed within the block, we would not be able to do correct type inference this way.

Reply:
Thanks for the explanation. I was wondering, because `srcs` and `dsts` for `block` are already occupied by the end of the block proxy,
```cpp
process(block, StackSignature{Type::none, block->params, false})
```
This might not work as is; but maybe we can think about how to fix it when we need it. But I think it’d be good to have some comments on `StackFlow` that if the expression is a block, this has some special meaning (the proxy for the end of block).

Reply:
Sure, I'll add some comments explaining how `block` is used in `srcs` and `dests` 👍 

> ```c++
> process(block, StackSignature{Type::none, block->params, false})
> ```
> 
> This might not work as is; but maybe we can think about how to fix it when we need it. 

Besides `block->params` not yet existing, I don't think there would be any issues with this.